database/table.go

```diff
@@ -103,22 +109,79 @@ func (t *Table) generateKey(d document.Document) ([]byte, error) {
 		return encoding.EncodeValue(v)
 	}
 
+	docid, err := t.generateDocid()
+	if err != nil {
+		return nil, err
+	}
+
+	return encoding.EncodeInt64(docid), nil
+}
+
+// this function looks up for the highest key in the table,
+// increments it, caches it in the database tableDocids map
+// and returns it.
+// if the docid is greater than max.Int64, it looks up for the lowest
+// available docid in the table.
+// Generating a docid is safe for concurrent access across
+// multiple transactions.
+func (t *Table) generateDocid() (int64, error) {
 	t.tx.db.mu.Lock()
 	defer t.tx.db.mu.Unlock()
 
-	cfg, err = t.cfgStore.Get(t.name)
-	if err != nil {
-		return nil, err
+	var err error
+
+	// get the cached latest docid
+	lastDocid, ok := t.tx.db.tableDocids[t.name]
+	// if no key was found in the cache, get the largest key in the table
+	if !ok {
+		it := t.Store.NewIterator(engine.IteratorConfig{Reverse: true})
+		it.Seek(nil)
+		if it.Valid() {
+			t.tx.db.tableDocids[t.name], err = encoding.DecodeInt64(it.Item().Key())
+			if err != nil {
+				return 0, err
+			}
+		} else {
+			t.tx.db.tableDocids[t.name] = 0
+		}
+
+		err = it.Close()
+		if err != nil {
+			return 0, err
+		}
+
+		lastDocid = t.tx.db.tableDocids[t.name]
 	}
 
-	cfg.LastKey++
-	key = encoding.EncodeInt64(cfg.LastKey)
-	err = t.cfgStore.Replace(t.name, cfg)
-	if err != nil {
-		return nil, err
+	// if the id is bigger than an int64
+	// look for the smallest available docid
+	if lastDocid > math.MaxInt64-1 {
+		return t.getSmallestAvailableDocid()
 	}
 
-	return key, nil
+	lastDocid++
+
+	// cache it
+	t.tx.db.tableDocids[t.name] = lastDocid
+	return lastDocid, nil
+}
+
+// getSmallestAvailableDocid iterates through the table store
+// and returns the first available docid.
+func (t *Table) getSmallestAvailableDocid() (int64, error) {
+	it := t.Store.NewIterator(engine.IteratorConfig{})
+	defer it.Close()
+
+	var i int64 = 1
+
+	for it.Seek(nil); it.Valid(); it.Next() {
+		if !bytes.Equal(it.Item().Key(), encoding.EncodeInt64(i)) {
+			return i, nil
+		}
+		i++
+	}
+
+	return 0, errors.New("reached maximum number of documents in a table")
```

Comment:
We can reach that point if `i == math.MaxInt64`?

Reply:
If ` == math.MaxInt64 + 1` yes